From d7b472cd88bfb58b75792f3e8c5ecbfe384a55d7 Mon Sep 17 00:00:00 2001
From: Sultan Alsawaf <sultan@kerneltoast.com>
Date: Tue, 3 May 2022 23:32:26 -0700
Subject: [PATCH] PM / wakeup: Avoid excessive s2idle wake attempts in
 pm_system_wakeup()

Calling s2idle_wake() once after pm_abort_suspend is incremented is all
that's needed to wake up from s2idle. Avoid multiple unnecessary attempts
to wake from s2idle by only doing the wakeup when pm_abort_suspend hits 1.
The s2idle machinery already provides the synchronization needed to make
this safe.

Signed-off-by: Sultan Alsawaf <sultan@kerneltoast.com>
---
 drivers/base/power/wakeup.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/base/power/wakeup.c b/drivers/base/power/wakeup.c
index 770bb04b7ed8b..093fad92e8e51 100644
--- a/drivers/base/power/wakeup.c
+++ b/drivers/base/power/wakeup.c
@@ -935,8 +935,8 @@ EXPORT_SYMBOL_GPL(pm_wakeup_pending);
 
 void pm_system_wakeup(void)
 {
-	atomic_inc(&pm_abort_suspend);
-	s2idle_wake();
+	if (atomic_inc_return_relaxed(&pm_abort_suspend) == 1)
+		s2idle_wake();
 }
 EXPORT_SYMBOL_GPL(pm_system_wakeup);
 
